### PR TITLE
Ensure diagnostics are set even for failed requests

### PR DESCRIPTION
**Pull request by abhijitpai**
## Description
Diagnostics were not set in the CosmosResponse for failed requests; set them.

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Util/Extensions.cs [hunk 1]
```diff
@@ -87,6 +87,16 @@ internal static ResponseMessage ToCosmosResponseMessage(this DocumentClientExcep
...
 
+            cosmosResponse.Diagnostics = new PointOperationStatistics(
+                activityId: cosmosResponse.Headers.ActivityId,
+                statusCode: dce.StatusCode.Value,
+                subStatusCode: SubStatusCodes.Unknown,
+                requestCharge: cosmosResponse.Headers.RequestCharge,
+                errorMessage: cosmosResponse.ErrorMessage,
+                method: request?.Method,
+                requestUri: request?.RequestUri,
+                clientSideRequestStatistics: null);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Util/Extensions.cs]: DocumentClientException  has a RequestStatistics property that can be used to set this value.

File: Microsoft.Azure.Cosmos/src/Util/Extensions.cs [hunk 2]
```diff
@@ -52,14 +52,14 @@ internal static ResponseMessage ToCosmosResponseMessage(this DocumentServiceResp
 
         internal static ResponseMessage ToCosmosResponseMessage(this DocumentClientException dce, RequestMessage request)
         {
-            // if StatusCode is null it is a client business logic error and it never hit the backend, so throw
-            if (dce.StatusCode == null)
+            // if StatusCode is not set, it is a client business logic error and it never hit the backend, so throw
+            if (!dce.StatusCode.HasValue)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Util/Extensions.cs]: Do not throw this. This is breaking change and will cause all the stream APIs to throw exceptions that were previously ResponseMessages.

Comment (abhijitpai) [reply Microsoft.Azure.Cosmos/src/Util/Extensions.cs]: I didn't change anything functionally here - I just made code clearer: StatusCode is not nullable but dce.StatusCode is a StatusCode?, and after checking that it is set, there is no meaning to defaulting to InternalServerError for the null case in line 62. I will revert this part of the change so there is no confusion, please fix it separately after considering impacts.

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Util/Extensions.cs]: You are correct. Sorry for the mistake. You didn't need to revert the change.

**Comment by kirankumarkolli** [comment]
@j82w can u please take a look and merge it?

**Pull request merged**